### PR TITLE
feat(wasm-gen): implement sys-call patterns

### PR DESCRIPTION
**Pull request by StackOverflowExcept1on**
Related to #3091
The PR is currently targeting another branch so as not to rewrite the code twice.

TODO:
- [x] implement `reply_from_reservation`
- [x] implement `gr_send_init -> gr_send_push -> gr_send_commit` pattern
- [x] implement more patterns (maybe cover `gr_send_commit_wgas`, `gr_send_push_input`)
- [x] don't repeat code and extract some stuff into methods
- [ ] consider `memcpy` function, some helper functions to create instructions

### CODE REVIEW

**Review by breathx — APPROVED** [review]

**Comment by StackOverflowExcept1on** [comment]
@CodiumAI-Agent /review

**Comment by StackOverflowExcept1on** [comment]
@CodiumAI-Agent /improve

File: utils/wasm-gen/src/generator/syscalls/imports.rs [hunk 1]
```diff
@@ -181,6 +181,25 @@ impl<'a, 'b> SysCallsImportsGenerator<'a, 'b> {
...
+            if let Err(err) = result {
+                match err {
+                    PreciseSysCallError::RequiredImports(err) => log::trace!("{err}"),
+                    PreciseSysCallError::Arbitrary(err) => return Err(err),
+                }
+            }
+        }
+
+        Ok(())
+    }
```
[the remark this answers is not in the document]
Comment (StackOverflowExcept1on) [reply utils/wasm-gen/src/generator/syscalls/imports.rs:201]: not needed

File: utils/wasm-gen/src/generator/syscalls/imports.rs [hunk 2]
```diff
@@ -414,8 +429,311 @@ impl<'a, 'b> SysCallsImportsGenerator<'a, 'b> {
...
+        self.generate_proper_sys_call_invocation(SYS_CALL, func_instructions);
+
+        Ok(())
+    }
+}
+
+/// The set of sys-calls that need to be imported to create precise sys-call.
+#[derive(thiserror::Error, Debug)]
+#[error("The following sys-calls must be imported: {0:?}")]
+pub struct RequiredSysCalls(&'static [SysCallName]);
```
Comment (SabaunT) [inline utils/wasm-gen/src/generator/syscalls/imports.rs:728]: better move upper type declarations

**Review by SabaunT — APPROVED** [review]

**Pull request merged**